src/readpool/mod.rs

```diff
@@ -0,0 +1,176 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod config;
+
+use futures::Future;
+use futures_cpupool as cpupool;
+
+pub use self::config::Config;
+
+pub struct WorkerThreadContext {
+    // Metrics collector to be added here
+}
+
+impl Clone for WorkerThreadContext {
+    fn clone(&self) -> WorkerThreadContext {
+        WorkerThreadContext {}
+    }
+}
+
+pub struct ReadPool {
+    pool_read_critical: cpupool::CpuPool,
+    pool_read_high: cpupool::CpuPool,
+    pool_read_normal: cpupool::CpuPool,
+    pool_read_low: cpupool::CpuPool,
+    context: WorkerThreadContext,
+}
+
+impl Clone for ReadPool {
+    fn clone(&self) -> ReadPool {
+        ReadPool {
+            pool_read_critical: self.pool_read_critical.clone(),
+            pool_read_high: self.pool_read_high.clone(),
+            pool_read_normal: self.pool_read_normal.clone(),
+            pool_read_low: self.pool_read_low.clone(),
+            context: self.context.clone(),
+        }
+    }
+}
+
+impl ReadPool {
+    pub fn new(config: &Config) -> ReadPool {
+        ReadPool {
+            pool_read_critical: cpupool::Builder::new()
+                .name_prefix("readpool-c")
+                .pool_size(config.read_critical_concurrency)
+                .stack_size(config.stack_size.0 as usize)
+                .create(),
+            pool_read_high: cpupool::Builder::new()
+                .name_prefix("readpool-h")
+                .pool_size(config.read_high_concurrency)
+                .stack_size(config.stack_size.0 as usize)
+                .create(),
+            pool_read_normal: cpupool::Builder::new()
+                .name_prefix("readpool-n")
+                .pool_size(config.read_normal_concurrency)
+                .stack_size(config.stack_size.0 as usize)
+                .create(),
+            pool_read_low: cpupool::Builder::new()
+                .name_prefix("readpool-l")
+                .pool_size(config.read_low_concurrency)
+                .stack_size(config.stack_size.0 as usize)
+                .create(),
+            context: WorkerThreadContext {},
+        }
+    }
+
+    #[inline]
+    fn get_pool_by_priority(&self, priority: Priority) -> &cpupool::CpuPool {
+        match priority {
+            Priority::ReadCritical => &self.pool_read_critical,
+            Priority::ReadHigh => &self.pool_read_high,
+            Priority::ReadNormal => &self.pool_read_normal,
+            Priority::ReadLow => &self.pool_read_low,
+        }
+    }
+
+    pub fn future_execute<F, R>(
+        &self,
+        priority: Priority,
+        feature_factory: F,
```

Comment:
What is the meaning of feature_factory?

Reply:
Oops. that's a typo. It should be `future_factory`. `future_factory: F` is a function that produce a future so that it is named as `_factory`.